pkg/template/processor_test.go

```diff
@@ -0,0 +1,604 @@
+		require.NoError(t, err)
+		err = p.Apply(objs, 2*time.Second)
+
+		require.NoError(t, err)
+		verifyRoleBinding(t, cl, namespace)
+
+		// when
+		objs, err = p.Process(templateContent(roleBindingObj, newUser), values)
+		require.NoError(t, err)
+		err = p.Apply(objs, 2*time.Second)
+
+		// then
+		require.NoError(t, err)
+		binding := verifyRoleBinding(t, cl, namespace)
+		require.Len(t, binding.Subjects, 2)
+	})
+
+	t.Run("should fail to create template object", func(t *testing.T) {
+		// given
+		cl := test.NewFakeClient(t)
+		cl.MockCreate = func(ctx context.Context, obj runtime.Object) error {
+			return errors.New("failed to create resource")
+		}
+
+		s := addToScheme(t)
+		namespace, commit, user := templateVars()
+
+		values := paramsKeyValues(namespace, commit, user)
+
+		p := template.NewProcessor(cl, s)
+
+		// when
+		objs, err := p.Process(templateContent(roleBindingObj), values)
+		require.NoError(t, err)
+		err = p.Apply(objs, 2*time.Second)
+
+		// then
+		require.Error(t, err)
+	})
+
+	t.Run("should fail to update template object", func(t *testing.T) {
+		// given
+		cl := test.NewFakeClient(t)
+		cl.MockUpdate = func(ctx context.Context, obj runtime.Object) error {
+			return errors.New("failed to update resource")
+		}
+
+		s := addToScheme(t)
+		namespace, commit, user := templateVars()
+
+		values := paramsKeyValues(namespace, commit, user)
+
+		p := template.NewProcessor(cl, s)
+		objs, err := p.Process(templateContent(roleBindingObj), values)
+		require.NoError(t, err)
+		err = p.Apply(objs, 2*time.Second)
+		require.NoError(t, err)
+
+		// when
+		objs, err = p.Process(templateContent(roleBindingObj), values)
+		require.NoError(t, err)
+		err = p.Apply(objs, 2*time.Second)
+
+		// then
+		assert.Error(t, err)
+	})
+}
+
+func addToScheme(t *testing.T) *runtime.Scheme {
+	s := scheme.Scheme
+	err := apis.AddToScheme(s)
+	require.NoError(t, err)
+	return s
+}
+
+func templateContent(objs ...string) []byte {
+	tmpl := templateMetadata
+	for _, obj := range objs {
+		tmpl += obj
+	}
+
+	return []byte(tmpl + templateParams)
+}
+
+func paramsKeyValues(namespace, commit, user string) map[string]string {
+	return map[string]string{
+		"PROJECT_NAME": namespace,
+		"COMMIT":       commit,
+		"USER_NAME":    user,
+	}
+}
+
+func templateVars() (string, string, string) {
+	return getNameWithTimestamp("namespace"), getNameWithTimestamp("sha"), getNameWithTimestamp("user")
+}
+
+func getNameWithTimestamp(prefix string) string {
+	return fmt.Sprintf("%s-%d", prefix, time.Now().UnixNano())
+}
+
+func verifyResource(t *testing.T, objKind schema.ObjectKind, vars ...string) {
+	require.IsType(t, &unstructured.Unstructured{}, objKind)
+	obj := objKind.(*unstructured.Unstructured)
+	objJson, err := obj.MarshalJSON()
+	require.NoError(t, err, "failed to marshal json from unstructured object")
+	for _, v := range vars {
+		assert.Contains(t, string(objJson), v, fmt.Sprintf("missing %s", v))
```

Comment:
Shouldn't we verify that the variables are written at the correct places? Here it "only" verifies that the whole object contains the variable values, but doesn't verify that are used/set correctly.
Apart from that, maybe it would be good that all variables are replaced - in other words, that the object doesn't contain any string matching `${...}`

Reply:
good point. I'll refactor the test to compare the whole output, instead

Reply:
done in 39a07f3